### PR TITLE
Iceberg metadata in Nessie

### LINKED ISSUE #3619: Store Iceberg Metadata in Nessie

### PR DESCRIPTION
**Pull request by snazy**
This PR connects the dots of the already existing base functionality to have "content attachments" and storing Iceberg Metadata in Nessie.

Fixes #3619

### CODE REVIEW

File: servers/iceberg-fixtures/src/test/java/org/projectnessie/iceberg/metadata/TestNessieIceberg.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    assertThat(asJsonNode(TableMetadataParser.toJson(fromNessie)))
+        .isEqualTo(asJsonNode(tableMetadataJson));
+    JsonNode nessie2 = toNessie(fromNessie);
+
+    assertThat(nessie2).isEqualTo(nessie);
+  }
+
+  static Stream<String> randomViewMetadata() {
+    return IntStream.range(0, 100)
+        .mapToObj(i -> NessieIceberg.randomViewMetadata(5))
```
Comment (dimas-b) [inline servers/iceberg-fixtures/src/test/java/org/projectnessie/iceberg/metadata/TestNessieIceberg.java:62]: Why do we randomize? Random test inputs unnerve me :) unless there's a way to reproduce with a known seed... If we want to test odd JSON structure, would it be preferable to actually list them out explicitly?

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java [hunk 2]
```diff
@@ -376,4 +418,88 @@ public void multiget() throws BaseNessieClientServerException {
                     .isEqualTo(IcebergTable.builder().from(tableB).id(content.getId()).build()))
         .doesNotContainKey(ContentKey.of("noexist"));
   }
+
+  static Stream<Arguments> icebergMetadataScenarios() {
+    return Stream.of(loadIcebergMetadataScenario("table-three-snapshots", 3));
+  }
+
+  private static Arguments loadIcebergMetadataScenario(String name, int count) {
```
Comment (dimas-b) [inline servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java:426]: This method is always called with the same args.

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java [hunk 3]
```diff
@@ -376,4 +418,88 @@ public void multiget() throws BaseNessieClientServerException {
...
+          .containsExactly(tuple(key, content, previousContent));
+
+      // Compare operation on HEAD commit with the committed operation
+      LogResponse log = getApi().getCommitLog().reference(committed).fetch(FetchOption.ALL).get();
+      assertThat(log)
+          .extracting(LogResponse::getLogEntries, list(LogEntry.class))
+          .element(0)
+          .extracting(LogEntry::getOperations, list(Operation.class))
+          .element(0)
+          // Clear content ID for comparison
```
Comment (dimas-b) [inline servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java:498]: Where is the ID cleared?

**Review by dimas-b — COMMENTED** [review]
Just a really superficial first pass review

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java (hunk 2 above)
Comment (snazy) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java:426]: Yea - only one "scenario" at the moment

File: servers/iceberg-fixtures/src/test/java/org/projectnessie/iceberg/metadata/TestNessieIceberg.java (hunk 1 above)
Comment (snazy) [reply servers/iceberg-fixtures/src/test/java/org/projectnessie/iceberg/metadata/TestNessieIceberg.java:62]: Updated to a `Random` w/ fixed seed.

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java (hunk 3 above)
Comment (snazy) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestContents.java]: Never :( (left-over)

File: servers/store/src/test/java/org/projectnessie/server/store/TestStoreWorker.java [hunk 4]
```diff
@@ -399,6 +412,164 @@ void testSerde(Map.Entry<ByteString, Content> entry) {
         .isEqualTo(entry.getKey());
   }
 
+  private static JsonNode loadJson(String resource) {
+    URL url = Thread.currentThread().getContextClassLoader().getResource(resource);
```
Comment (dimas-b) [inline servers/store/src/test/java/org/projectnessie/server/store/TestStoreWorker.java:416]: Why `getContextClassLoader` and not `TestStoreWorker.class.getClassLoader()`? I suppose the latter is a more stable representation of the generated resources. It maps to the test class path, which maps to module dependencies,  which include the resources.

File: servers/store/src/test/java/org/projectnessie/server/store/TestStoreWorker.java [hunk 5]
```diff
@@ -399,6 +412,164 @@ void testSerde(Map.Entry<ByteString, Content> entry) {
...
+
+    assertThat(parsed.getIcebergRefState().getCurrentPartsList())
+        .containsExactlyInAnyOrderElementsOf(expectedStoreOnRef.getCurrentPartsList());
+    assertThat(parsed.getIcebergRefState().getMorePartsList())
+        .containsExactlyInAnyOrderElementsOf(expectedStoreOnRef.getMorePartsList());
+
+    assertThat(parsed.getIcebergRefState().toBuilder().clearCurrentParts().clearMoreParts().build())
+        .isEqualTo(expectedStoreOnRef.clearCurrentParts().clearMoreParts().build());
+
+    // TODO validate that metadata is really shallow (no child object arrays)
```
Comment (dimas-b) [inline servers/store/src/test/java/org/projectnessie/server/store/TestStoreWorker.java:489]: Still TODO?

File: model/src/main/java/org/projectnessie/model/IcebergView.java [hunk 6]
```diff
@@ -91,4 +92,29 @@ public static IcebergView of(
         .sqlText(sqlText)
         .build();
   }
+
+  public static IcebergView of(JsonNode metadata, String metadataLocation, String id) {
+    int currentVersionId = metadata.get(IcebergContent.CURRENT_VERSION_ID).asInt(-1);
+    String sqlText = "";
+    String dialect = ""; // TODO !!
```
[the remark this answers is not in the document]
Comment (snazy) [reply model/src/main/java/org/projectnessie/model/IcebergView.java:99]: Sadly yes, waiting for the formal Iceberg Views spec. /cc @nastra

File: model/src/main/java/org/projectnessie/model/IcebergView.java [hunk 7]
```diff
@@ -91,4 +92,29 @@ public static IcebergView of(
...
+    int schemaId = 0;
+    for (JsonNode version : metadata.get(IcebergContent.VERSIONS)) {
+      if (version.get(IcebergContent.VERSION_ID).asInt(-1) == currentVersionId) {
```
[the remark this answers is not in the document]
Comment (snazy) [reply model/src/main/java/org/projectnessie/model/IcebergView.java:102]: Not sure whether it should be the last one. If it's not set, then there shouldn't be any versions at all IMO. Maybe @nastra has an opionion on this.

[the remark this answers is not in the document]
Comment (nastra) [reply model/src/main/java/org/projectnessie/model/IcebergView.java:102]: code makes sense to me

File: servers/iceberg-fixtures/src/test/java/org/projectnessie/iceberg/metadata/TestBigSchema.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    byte[] compressedIceberg = compress(jsonIceberg);
+    byte[] compressedNessie = compress(jsonNessie);
+
+    System.err.printf(
+        "  num-cols: %-8d   iceberg-json-size: %10d / %-10d   nessie-json-size: %10d / %-10d%n",
+        numColumns,
+        jsonIceberg.length(),
+        compressedIceberg.length,
+        jsonNessie.length(),
+        compressedNessie.length);
```
[the remark this answers is not in the document]
Comment (snazy) [reply servers/iceberg-fixtures/src/test/java/org/projectnessie/iceberg/metadata/TestBigSchema.java:61]: nope

File: servers/iceberg-fixtures/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  testImplementation("org.assertj:assertj-core")
+  testImplementation("org.junit.jupiter:junit-jupiter-api")
+  testImplementation("org.junit.jupiter:junit-jupiter-engine")
+  testImplementation("org.junit.jupiter:junit-jupiter-params")
+
+  testCompileOnly(platform("io.quarkus:quarkus-bom"))
+  testCompileOnly("org.eclipse.microprofile.openapi:microprofile-openapi-api")
+}
+
+val generatedIcebergMetadataDir = project.buildDir.resolve("generated/iceberg")
```
[the remark this answers is not in the document]
Comment (snazy) [reply servers/iceberg-fixtures/build.gradle.kts:63]: fixed.
Note: the name of the directory doesn't matter, but the directory was missing as a resource directory.

File: servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+    fileWriter.writeFile("table-three-snapshots", "2", TableMetadataParser.toJson(metadata));
+
+    Snapshot snapshot3 = randomSnapshot(metadata);
+    metadata = TableMetadata.buildFrom(metadata).setCurrentSnapshot(snapshot3).build();
+
+    fileWriter.writeFile("table-three-snapshots", "3", TableMetadataParser.toJson(metadata));
+  }
+
+  private static void cleanTargetDir(Path targetBaseDir) throws IOException {
```
[the remark this answers is not in the document]
Comment (snazy) [reply servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java:139]: npoe

File: servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+    Snapshot snapshot2 = randomSnapshot(metadata);
+    metadata = TableMetadata.buildFrom(metadata).setCurrentSnapshot(snapshot2).build();
+
+    fileWriter.writeFile("table-three-snapshots", "2", TableMetadataParser.toJson(metadata));
+
+    Snapshot snapshot3 = randomSnapshot(metadata);
+    metadata = TableMetadata.buildFrom(metadata).setCurrentSnapshot(snapshot3).build();
+
+    fileWriter.writeFile("table-three-snapshots", "3", TableMetadataParser.toJson(metadata));
```
[the remark this answers is not in the document]
Comment (snazy) [reply servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java:136]: there are 3 snapshots for that scenario. all "scenarios" get their own directory.

File: servers/store/src/main/java/org/projectnessie/server/store/TableCommitMetaStoreWorker.java [hunk 12]
```diff
@@ -109,9 +168,113 @@ private static void toStoreIcebergTable(IcebergTable table, ObjectTypes.Content.
...
+                def.contentPartType,
+                id,
+                jsonNode,
+                attachmentConsumer,
+                id == currentId ? currentPartRefConsumer : morePartRefConsumer);
+          }
+        });
+
+    // TODO Once we only return the relevant child objects (current snapshot, schema, etc), we can
+    //  clear TableMetadata.snapshotLog and TableMetadata.previousFiles.
```
[the remark this answers is not in the document]
Comment (snazy) [reply servers/store/src/main/java/org/projectnessie/server/store/TableCommitMetaStoreWorker.java:237]: Wanna keep the behavior as it currently is with on-reference-state - to not accidentally break anything. Not yet 100% sure which parts we can and cannot omit.

File: servers/iceberg-fixtures/src/main/java/org/apache/iceberg/view/IcebergBride.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+ */
+package org.apache.iceberg.view;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.io.StringWriter;
+import org.apache.iceberg.util.JsonUtil;
+
+public final class IcebergBride {
```
[the remark this answers is not in the document]
Comment (snazy) [reply servers/iceberg-fixtures/src/main/java/org/apache/iceberg/view/IcebergBride.java:24]: Yea - the whole fixtures module is a test-only thing for other modules in servers. Not used anywhere else.

File: servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java:136]: right, I mean that file names like `3.json` are a bit too terse :) How about `table-three-snapshots-N.json`?

File: servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/NessieIceberg.java [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+    Types.BooleanType.get(),
+    Types.DateType.get(),
+    Types.TimeType.get(),
+    Types.UUIDType.get(),
+    Types.TimestampType.withZone(),
+    Types.TimestampType.withoutZone()
+  };
+
+  // Deterministic w/ constant seed
+  static final Random RANDOM = new Random(42L);
```
Comment (dimas-b) [inline servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/NessieIceberg.java:81]: Well, it's still reused across invocations of the various data-generating methods so each generation is still volatile based on the total order of tests :thinking: WDYT about using a fresh fixed-seed random for each top-level generation method call?

File: servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (snazy) [reply servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java:139]: To have a clean output when things change.

File: servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (snazy) [reply servers/iceberg-fixtures/src/main/java/org/projectnessie/iceberg/metadata/GenerateIcebergFiles.java:136]: updated

File: model/src/main/java/org/projectnessie/model/IcebergView.java [hunk 15]
```diff
@@ -91,4 +92,35 @@ public static IcebergView of(
...
+  public static IcebergView of(JsonNode metadata, String metadataLocation, String id) {
+    int currentVersionId = metadata.path(CURRENT_VERSION_ID).requireNonNull().asInt(-1);
+
+    String sqlText = "";
+    String dialect = ""; // TODO dialect is currently undefined in Iceberg
+    int schemaId = 0;
+    JsonNode versions = metadata.get(VERSIONS);
+    if (versions != null) {
+      for (JsonNode version : versions) {
+        int versionId = version.path(VERSION_ID).requireNonNull().asInt();
```
[the remark this answers is not in the document]
Comment (snazy) [reply model/src/main/java/org/projectnessie/model/IcebergView.java:105]: requireNotNull requires that VERSION_ID must be explicitly set

**Review by dimas-b — APPROVED** [review]

**Comment by snazy** [comment]
Ran these changes in the WIP of the tools-integration-tests, and those unveiled some issues, which are fixed by two new commits.

Also had to remove the metadata generator code, because it would have introduced a non-fixable dependency issue wrt the included builds pattern of the tools-integration-tests.

**Review by dimas-b — APPROVED** [review]

**Comment by snazy** [comment]
FYI integrations-tests run of this change against latest Iceberg [passed](https://github.com/projectnessie/nessie/runs/7317001344?check_suite_focus=true)

**Pull request merged**